presto-main/src/main/java/com/facebook/presto/operator/TaskContext.java

```diff
@@ -90,6 +92,8 @@
     @GuardedBy("cumulativeMemoryLock")
     private long lastTaskStatCallNanos = 0;
 
+    private final AtomicBoolean initialized = new AtomicBoolean();
```

Comment:
Does it need to be atomic?

Reply:
AFAIK multiple threads can be accessing instances of this class, also looking at the other fields and the `@ThreadSafe` annotation I think we should make this an atomic boolean.

Reply:
I understand, but..

1. is it, in fact, any more thread-safe with this being `AtomicBoolean`? Do you intend to use unsafe publication to pass `TaskContext` to another thread after `initialize()` was called? If we assume `initialize()` happens always just after `new`, then in practise atomicity seems redundant. I view this flag as only safety-check that `initialize()` wasn't forgotten by a programmer, rather then a run-time control. (However, even if it is for formalism of `@ThreadSafe`-ty, I understand that.)
2. using non-atomic is usually faster, as it can be cached

Reply:
What @findepi says.  It seems that `initialize()` is invoked immediately after construction and at most once.  You could just use a volatile boolean.  `initialized` is written only once and all its reads will have the correct value since `volatile`'s aren't cached either.

Reply:
@lvijay i didn't mean `volatile` -- see my point 2.

Reply:
>> If we assume initialize() happens always just after new, then in practise atomicity seems redundant

Yes, this is the case for now. The same thread calls `initialize()` after `new`, so atomicity in the code's current form is redundant. But, this is only true for now as we don't know how future clients of this class will use it.

Also, the `initialized` flag is read by other methods in this class (the `checkState()` calls) called by other threads (e.g., `start()` will be called by the driver/worker threads), so we need visibility for this variable.

Reply:
@nezihyigitbasi , i agree: _technically_ something is needed to guarantee visibility and plain `boolean` is not enough.
My only concern is that _practically_ strong visibility guarantee is not needed for this member, as its only function is to prevent coding mistake (forgetting about `initialize()`).

I am OK either way.